File: src/lab.cpp
/**
 * @brief 并查集
 * 0-index
 */
#include <algorithm>
#include <iostream>
#include <map>
#include <unordered_set>
#include <vector>
using namespace std;
using i64         = long long;
using PII         = pair<int, int>;
constexpr int INF = 1e9;

struct DSU {
    int              component;
    std::vector<int> f, siz;

    DSU() {}

    DSU(int n) : component(n)
    {
        init(n);
    }

    void init(int n)
    {
        f.resize(n);
        std::iota(f.begin(), f.end(), 0);
        siz.assign(n, 1);
    }

    int find(int x)
    {
        while (x != f[x]) {
            x = f[x] = f[f[x]];
        }
        return x;
    }

    bool same(int x, int y)
    {
        return find(x) == find(y);
    }

    bool merge(int x, int y)
    {
        x = find(x);
        y = find(y);
        if (x == y) {
            return false;
        }
        siz[x] += siz[y];
        f[y] = x;
        component--;
        return true;
    }

    int size(int x)
    {
        return siz[find(x)];
    }

    std::vector<std::vector<int>> getGroups(void)
    {
        std::vector<std::vector<int>>   res;
        std::map<int, std::vector<int>> mp;
        for (int i = 0; i < f.size(); i++) {
            mp[find(i)].emplace_back(i);
        }
        res.reserve(mp.size());
        for (auto& [_, group] : mp) {
            res.emplace_back(std::move(group));
        }
        return res;
    }
};

vector<int> Eulerian(int n, vector<unordered_set<int>> graph, vector<int> deg)
{
    int s = -1, t = -1, cnt = 0;
    for (int i = 1; i <= n; i++) {
        if (deg[i] % 2 == 0) {
            cnt++;
        } else if (s == -1) {
            s = i;
        } else if (t == -1) {
            t = i;
        }
    }
    vector<int> ans;
    auto        dfs = [&](auto self, int u) -> void {
        while (graph[u].size()) {
            int v = *graph[u].begin();
            graph[u].erase(graph[u].find(v));
            graph[v].erase(graph[v].find(u));
            self(self, v);
        }
        ans.push_back(u);
    };
    dfs(dfs, s);
    reverse(ans.begin(), ans.end());
    return ans;
}

signed main(int argc, char** argv)
{
    ios::sync_with_stdio(false);
    cin.tie(0);
    cout.tie(0);
#ifdef BATCH
    freopen(argv[1], "r", stdin);
    freopen(argv[2], "w", stdout);
#endif

    return 0;
}

File: src/lab.py
import math
import sys
input = sys.stdin.read

def init():
    return int(input())

def print(x):
    if x < 0:
        x = -x
        sys.stdout.write('-')
    if x // 10:
        print(x // 10)
    sys.stdout.write(str(x % 10))

N = 100 + 5
n, X = init()
p = [0] * N
dp = [[float('inf')] * N for _ in range(N)]
num = [0] * N

s = []
for _ in range(n):
    s.append((init(), init()))

def mn(x, y):
    return x if x < y else y

Ans = float('inf')

def cmp(x, y, z):
    return math.log2(x) + z * math.log2(y) >= math.log2(X)

def quick_mod(x, y):
    s = 1
    while y:
        if y & 1:
            s *= x
        x *= x
        y >>= 1
    return s

def dfs(i, mul, calc):
    global Ans
    if calc >= Ans:
        return
    if i > p[0]:
        if mul >= X:
            Ans = mn(Ans, calc)
        return
    for cnt in range(0, num[i] + 1, 2):
        if cnt == 0:
            dfs(i + 1, mul, calc)
            continue
        if cmp(mul, p[i], cnt):
            Ans = mn(Ans, calc + dp[i][cnt])
            return
        dfs(i + 1, mul * quick_mod(p[i], cnt), calc + dp[i][cnt])

p[0] = 0
for i in range(1, n + 1):
    who = 0
    for j in range(1, p[0] + 1):
        if p[j] == s[i - 1][0]:
            who = j
    if who == 0:
        p[0] += 1
        p[p[0]] = s[i - 1][0]
        dp[p[0]][1] = s[i - 1][1]
        num[p[0]] = 1
    else:
        num[who] += 1
        for j in range(num[who], 0, -1):
            dp[who][j] = mn(dp[who][j], dp[who][j - 1] + s[i - 1][1])

dfs(1, 1, 0)
if Ans == float('inf'):
    print("-1")
else:
    print(Ans)
